src/stan/lang/grammars/semantic_actions_def.cpp

```diff
@@ -3124,90 +3055,221 @@ namespace stan {
+      var_decl_result = result;
+    }
+    boost::phoenix::function<validate_array_local_var_decl>
+    validate_array_local_var_decl_f;
+
+    void validate_fun_arg_var::operator()(var_decl& var_decl_result,
+                                      const bare_expr_type& bare_type,
+                                      const std::string& name,
+                                      bool& pass,
+                                      std::ostream& error_msgs) const {
+      if (bare_type.is_ill_formed_type()) {
+        error_msgs << "Function argument is ill formed,"
+                   << " name="
+                   << name
+                   << std::endl;
+        pass = false;
+        return;
+      }
+      stan::lang::var_decl vd(name, bare_type);
+      var_decl_result = vd;
+    }
+    boost::phoenix::function<validate_fun_arg_var>
+    validate_fun_arg_var_f;
+
+    void validate_bare_type_is_data::operator()(
+                                       bare_expr_type& bare_type,
+                                       scope& scope,
+                                       bool& pass,
+                                       std::ostream& error_msgs) const {
+      if (bare_type.is_int_type()) {
+          pass = false;
+          error_msgs
+            << "Data qualifier cannot be applied to variable of type int"
+            << std::endl;
+          return;
+      }
+      if (scope.program_block() == data_origin)
+        bare_type.set_is_data();
+    }
+    boost::phoenix::function<validate_bare_type_is_data>
+    validate_bare_type_is_data_f;
+
+    void validate_bare_type::operator()(
+                                        bare_expr_type& bare_type_result,
+                                        const bare_expr_type& el_type,
+                                        const size_t& num_dims,
+                                        bool& pass,
+                                        std::ostream& error_msgs) const {
+      if (el_type.is_ill_formed_type()) {
+        error_msgs << "Ill-formed bare type" << std::endl;
         pass = false;
-        error_msgs << "duplicate declaration of variable, name="
-                   << var_decl.name_;
+        return;
+      }
+      pass = true;
+      if (num_dims == 0) {
+        bare_type_result = el_type;
+        return;
+      }
+      stan::lang::bare_array_type bat(el_type);
+      for (size_t i=0; i < num_dims-1; ++i) {
+        stan::lang::bare_expr_type cur_type(bat);
+        bat = bare_array_type(cur_type);
+      }
+      bare_type_result = bat;
+    }
+    boost::phoenix::function<validate_bare_type>
+    validate_bare_type_f;
 
-        error_msgs << "; attempt to redeclare as ";
+    template <typename T>
+    void add_to_var_map::operator()(const T& decl,
+                                   variable_map& vm, bool& pass,
+                                   const scope& var_scope,
+                                   std::ostream& error_msgs) const {
+      pass = false;
```

Comment:
This initial pass = false isn't necessary as every logic branch now sets it.  I think that's an improvement over how it was.